### ISSUE TITLE
any approach to override icon/renderer for specific folder?

### ISSUE THREAD

**Issue by linlol** [jupyterlab/jupyterlab#15741]
### Problem

We are working on an internal jupyter lab solution. In this solution, there is some folder managed by ourselves (for example, a folder controlled by git records common util notebooks)

Is there any approach to specify another piece of svg string to notebooks with special name pattern? 

This could also be a common requirement I guess? Suppose the case that give node_modules a dedicated icon 

### Proposed Solution

One possible solution is leave an abstraction like fileType, which enables application maintainer to customise it in lab extension

With such kind of abstraction, it shall also be better if we can make it configurable in setting registry

**Comment by krassowski** [comment]
This is already possible. You can register a custom file type ([`IFileType`](https://jupyterlab.readthedocs.io/en/latest/api/interfaces/rendermime_interfaces.IRenderMime.IFileType.html)) on the document registry and provide a custom icon. You can do it directly by calling [`DocumentRegistry.addFileType`](https://jupyterlab.readthedocs.io/en/latest/api/classes/docregistry.DocumentRegistry-1.html#addFileType) which would be likely the best approach for you, or via a MIME extension see [this example](https://github.com/jupyterlab/extension-examples/tree/main/mimerenderer). Note that either way you would provide an empty list for `extensions` argument and set the optional [`IFileType.pattern`](https://jupyterlab.readthedocs.io/en/latest/api/interfaces/rendermime_interfaces.IRenderMime.IFileType.html#pattern), e.g. to `^node_modules$`.

**Comment by linlol** [comment]
thanks @krassowski !

Yes I found this yeaterday and blocked by the mimetype setting.

Would have a try base on your example, thanks a lot!

**Comment by linlol** [comment]
Hi @krassowski 

I have a try as below and somehow it doesn't work even though I can see my customised views registered in document manager

This is what I tried (I am sorry that I cannot share screenshot/source code of internal project)

1. Setup regular extension typed as **jupyterfrontendplugin**<void> with addfiletype logic called in activate function

2. Create customised fileType from **DocumentRegistry.getDefaultDirectoryFileType** function, and override name/display name as dummy, customise pattern as "^node_modules$", icon customised as ui-components/JuliaIcon as example

3. get DocumentRegistry instance from global jupyerFrontEnd instance and call **docRegistry.addFileType**

After above operations, I can see my customised fileType registered in setting editor, but icon still not changed. Maybe it is overriden by default directory rule, is there anything like rank?

**Comment by krassowski** [comment]
Directories and notebooks are treated specially:

https://github.com/jupyterlab/jupyterlab/blob/82051a9d0c6dc93677b0b66560aea48d71aaa043/packages/docregistry/src/registry.ts#L664-L692

Maybe we need to make a small change in the `docregistry` code? Would you be open to submitting a pull request?

**Comment by linlol** [comment]
Just had a full-look at related code, looks like notebook and directory would be redirected to default file types and my specified pattern would be skipped.

I may propose below logic

1. For notebook/directory apply fileType by path check via logic like

**const tryFts = this.getFileTypesForPath(name)**

2. If we get non-null fileType, we check if tryFts.contentType is directory (or notebook respectively)

If so, use user-customised fileType

If not, fallback to the default notebook/directory fileType.

If this proposal makes sense to you, I would try my best to propose this change this month (It would not be that quick as I would not be able to access my personal PC for around 2 weeks due to personal issue)

**Comment by linlol** [comment]
Just had a full-look, looks like notebook and directory would be forced to default file types such that pattern matching logic  would be skipped.


I may propose below logic update

1. For notebook/directory apply fileType by path check via logic like

**const tryFts = this.getFileTypesForPath(name)**

2. If we get non-null fileType, we check if tryFts.contentType is directory (or notebook respectively)

If so, use user-customised fileType

If not, fallback to the default notebook/directory fileType.



@krassowski please take a glance if it sense to you, if so, I would try to make it this mont

 (It would not be that quick as I would not be able to access my personal PC for around 2 weeks due to personal issue)

**Comment by JasonWeill** [comment]
@linlol Thank you for opening this issue! Which version of JupyterLab are you using?

**Comment by linlol** [comment]
> @linlol Thank you for opening this issue! Which version of JupyterLab are you using?

Hi @JasonWeill , I am currently using 3.6.3

As you can see from the discussion among Krassowski and I, the behaviour shall be indifferent for the latest dev branch

**Comment by JasonWeill** [comment]
@linlol Thanks for your response. If possible, I recommend upgrading to Lab 4. Most new features are not getting backported to Lab 3.

**Comment by linlol** [comment]
> @linlol Thanks for your response. If possible, I recommend upgrading to Lab 4. Most new features are not getting backported to Lab 3.

@JasonWeill yes, thanks for pointing this, yes definitely we would upgrade jupyter lab 4

**Comment by linlol** [comment]
> > @linlol Thanks for your response. If possible, I recommend upgrading to Lab 4. Most new features are not getting backported to Lab 3.
> 
> @JasonWeill yes, thanks for pointing this, yes definitely we would upgrade jupyter lab 4

definitely I would not expect this feature onboarding at 3.6.x🤣🤣

**Comment by krassowski** [comment]
The logic change that you propose appears fine to me - feel welcome to start a PR, from there we can see if this breaks any tests etc. That would be a small behaviour change so I believe we could introduce it in JupyterLab 4.2.0.

**Comment by linlol** [comment]
thanks, I would start making the change among the end of the month (currently I don't have eligible coding env due to national holiday

**Comment by linlol** [comment]
hi @krassowski , I make a PR for my change, is there any easy approach to propose a quick test
